### PR TITLE
[Validator] Fix deprecation for ctype_* functions

### PR DESCRIPTION
**Pull request by adpeyre**
| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      |  no
| New feature?  | no
| Deprecations? | yes
| Issues        | 
| License       | MIT

`Deprecated: ctype_digit(): Argument of type int will be interpreted as string in the future`

Fix deprecation for ctype_* functions used in [Type Constraint](https://symfony.com/doc/current/reference/constraints/Type.html#type-1).

`As of PHP 8.1.0, passing a non-string argument is deprecated. In the future, the argument will be interpreted as a string instead of an ASCII codepoint. Depending on the intended behavior, the argument should either be cast to [string](https://www.php.net/manual/en/language.types.string.php) or an explicit call to [chr()](https://www.php.net/manual/en/function.chr.php) should be made.`

Because it's not a bug but a minor deprecation, I've started a branch from 6.4.

### CODE REVIEW

**Comment by xabbuh** [comment]
Can you explain which deprecation you are trying to fix here?

File: src/Symfony/Component/Validator/Constraints/TypeValidator.php [hunk 1]
```diff
@@ -92,4 +88,32 @@ public function validate(mixed $value, Constraint $constraint)
             ->setCode(Type::INVALID_TYPE_ERROR)
             ->addViolation();
     }
+
+
+    private function isValidByFunction(string $type, mixed $value): bool
+    {
+        $validationFunction = self::VALIDATION_FUNCTIONS[$type] ?? null;
+
+        if (null === $validationFunction) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/TypeValidator.php:97]:
```suggestion
        if (null === $validationFunction = self::VALIDATION_FUNCTIONS[$type] ?? null) {
```

File: src/Symfony/Component/Validator/Constraints/TypeValidator.php [hunk 2]
```diff
@@ -92,4 +88,32 @@ public function validate(mixed $value, Constraint $constraint)
...
+        if (!match ($type) {
+            'finite-float' => \is_float($value) && is_finite($value),
+            'finite-number' => \is_int($value) || \is_float($value) && is_finite($value),
+            'number' => \is_int($value) || \is_float($value) && !is_nan($value),
+            default => $validationFunction($value),
+        }) {
+            return false;
+        }
+
+        return true;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/TypeValidator.php:117]:
```suggestion
        return match ($type) {
            'finite-float' => \is_float($value) && is_finite($value),
            'finite-number' => \is_int($value) || \is_float($value) && is_finite($value),
            'number' => \is_int($value) || \is_float($value) && !is_nan($value),
            default => (\is_string($value) || !str_starts_with($validationFunction, 'ctype_')) && $validationFunction($value),
        };
```

**Review by nicolas-grekas — COMMENTED** [review]
Does a similar code exist in 5.4? If yes, please target 5.4 instead as we consider those changes as bug fixes.

**Comment by nicolas-grekas** [comment]
Here is the deprecation since PHP 8.4:

> var_dump(ctype_digit(123));
> Deprecated: ctype_digit(): Argument of type int will be interpreted as string in the future in php shell code on line 1
bool(false)

**Comment by adpeyre** [comment]
Since PHP 8.1 @nicolas-grekas ;)

Edit : ok, the message will change in 8.4.

**Pull request closed without merging**